### PR TITLE
Throw converter not found instead of generic IndexOutOfBound exception

### PR DESCRIPTION
**Pull request by ar0d10n**
Throw converter not found instead of generic IndexOutOfBound exception

## Motivation and Context

#3845 

## Modifications

Added parameter validation and throwing appropriate exception instead of generic IndexOutOfBound for default attribute converter and enhanced document

## Testing

Added unit tests 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/document/DefaultEnhancedDocument.java [hunk 1]
```diff
@@ -92,13 +92,21 @@ public static <T> AttributeConverter<T> converterForClass(EnhancedType<T> type,
...
         if (type.rawClass().isAssignableFrom(Map.class)) {
+            List<EnhancedType<?>> params = type.rawClassParameters();
+            if (params.isEmpty()) {
+                throw new IllegalStateException("Type parameters are missing for " + type);
+            }
             return (AttributeConverter<T>) MapAttributeConverter.mapConverter(
-                StringConverterProvider.defaultProvider().converterFor(type.rawClassParameters().get(0)),
-                converterForClass(type.rawClassParameters().get(1), chainConverterProvider));
+                StringConverterProvider.defaultProvider().converterFor(params.get(0)),
+                converterForClass(params.get(1), chainConverterProvider));
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/document/DefaultEnhancedDocument.java:109]: Is it possible to end up in a state where the rawClassParameters only contains one element? Should we check that it has 2 entries?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/TableSchemaTest.java [hunk 2]
```diff
@@ -85,6 +87,14 @@ public void fromClass_constructsImmutableTableSchema() {
         assertThat(tableSchema).isInstanceOf(ImmutableTableSchema.class);
     }
 
+    @Test
+    public void fromClass_missingObjectConverter() {
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/TableSchemaTest.java:91]: Nit: Should contain the effect, maybe `fromClass_missingObjectConverter_throwsIllegalStateException`

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/TableSchemaTest.java [hunk 3]
```diff
@@ -85,6 +87,14 @@ public void fromClass_constructsImmutableTableSchema() {
...
+        assertThatIllegalStateException().isThrownBy(
+            () -> TableSchema.fromClass(ObjectMapBean.class)
+        ).withMessage("Converter not found for "
+                      + "EnhancedType(java.lang.Object)");
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/TableSchemaTest.java:95]: NitNit: Line break seems unnecessary, lines are pretty short?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/document/EnhancedDocumentTest.java [hunk 4]
```diff
@@ -304,6 +305,34 @@ void access_CustomType_without_AttributeConverterProvider() {
         assertThat(docWithCustomProvider.get("customMapValue", EnhancedType.of(CustomClassForDocumentAPI.class))).isNotNull();
     }
 
+    @Test
+    void error_When_listUsed() {
```
Comment (cenedhryn) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/document/EnhancedDocumentTest.java:309]: Nit: The test case names in this class don't seem to follow a rule, but can we use something like
`listUsed_missingObjectConverter_throwsIllegalStateException`?

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/document/DefaultEnhancedDocument.java (hunk 1 above)
Comment (ar0d10n) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/document/DefaultEnhancedDocument.java:109]: I think client code can do that, so I will add check for size

**Comment by ar0d10n** [comment]
closed in favour of #5538

**Pull request closed without merging**